File: libs/power-log-parser/src/lib/parsers/battlegrounds-quest-reward-equipped-parser.ts
import { CardType, GameTag, Zone } from '@firestone-hs/reference-data';
import { ActionParser } from '../action-parser';
import { GameEventHelper, GameEventProvider } from '../game-event';
import { FullEntity, Node, TagChange } from '../models';
import { GameState } from '../state/game-state';
import { ParserState, StateType } from '../state/parser-state';
import type { StateFacade } from '../state/state-facade';

export class BattlegroundsQuestRewardEquippedParser implements ActionParser {
	readonly ParserName = 'BattlegroundsQuestRewardEquippedParser';

	private GameState: GameState;
	private ParserState: ParserState;
	private StateFacade: StateFacade;

	constructor(parserState: ParserState, facade: StateFacade) {
		this.ParserState = parserState;
		this.GameState = parserState.GameState;
		this.StateFacade = facade;
	}

	AppliesOnNewNode(node: Node, stateType: StateType): boolean {
		return (
			stateType === StateType.PowerTaskList &&
			node.Type === TagChange &&
			(node.Object as TagChange).Name === (GameTag.ZONE as number) &&
			(node.Object as TagChange).Value === (Zone.PLAY as number) &&
			this.GameState.CurrentEntities.has((node.Object as TagChange).Entity) &&
			this.GameState.CurrentEntities.get((node.Object as TagChange).Entity)!.GetTag(
				GameTag.CARDTYPE,
			) === (CardType.BATTLEGROUND_QUEST_REWARD as number)
		);
	}

	AppliesOnCloseNode(node: Node, stateType: StateType): boolean {
		return (
			stateType === StateType.PowerTaskList &&
			node.Type === FullEntity &&
			(node.Object as FullEntity).GetTag(GameTag.ZONE) === (Zone.PLAY as number) &&
			(node.Object as FullEntity).GetTag(GameTag.CARDTYPE) ===
				(CardType.BATTLEGROUND_QUEST_REWARD as number)
		);
	}

	CreateGameEventProviderFromNew(node: Node): GameEventProvider[] | null {
		const tagChange = node.Object as TagChange;
		const entity = this.GameState.CurrentEntities.get(tagChange.Entity);
		if (entity == null) {
			return null;
		}

		const cardId = entity.CardId;
		const controllerId = entity.GetEffectiveController();
		const isHeroPowerReward = entity.GetTag(GameTag.BACON_IS_HEROPOWER_QUESTREWARD) === 1;
		return [
			GameEventProvider.Create(
				tagChange.TimeStamp,
				'BATTLEGROUNDS_QUEST_REWARD_EQUIPPED',
				GameEventHelper.CreateProvider(
					'BATTLEGROUNDS_QUEST_REWARD_EQUIPPED',
					cardId,
					controllerId,
					entity.Id,
					this.StateFacade,
					{
						PlayerId: entity.GetTag(GameTag.PLAYER_ID),
						IsHeroPowerReward: isHeroPowerReward,
					},
				),
				true,
				node,
			),
		];
	}

	CreateGameEventProviderFromClose(node: Node): GameEventProvider[] | null {
		const fullEntity = node.Object as FullEntity;
		const cardId = fullEntity.CardId;
		const controllerId = fullEntity.GetEffectiveController();
		const isHeroPowerReward = fullEntity.GetTag(GameTag.BACON_IS_HEROPOWER_QUESTREWARD) === 1;
		return [
			GameEventProvider.Create(
				fullEntity.TimeStamp,
				'BATTLEGROUNDS_QUEST_REWARD_EQUIPPED',
				GameEventHelper.CreateProvider(
					'BATTLEGROUNDS_QUEST_REWARD_EQUIPPED',
					cardId,
					controllerId,
					fullEntity.Id,
					this.StateFacade,
					{
						PlayerId: fullEntity.GetTag(GameTag.PLAYER_ID),
						IsHeroPowerReward: isHeroPowerReward,
					},
				),
				true,
				node,
			),
		];
	}
}

File: tools/linux-helpers/hs-converter/HearthstoneReplays/Events/Parsers/BattlegroundsQuestRewardEquippedParser.cs
using HearthstoneReplays.Parser;
using HearthstoneReplays.Enums;
using HearthstoneReplays.Parser.ReplayData.Entities;
using HearthstoneReplays.Parser.ReplayData.GameActions;
using System.Collections.Generic;

namespace HearthstoneReplays.Events.Parsers
{
    public class BattlegroundsQuestRewardEquippedParser : ActionParser
    {
        private GameState GameState { get; set; }
        private ParserState ParserState { get; set; }
        private StateFacade StateFacade { get; set; }

        public BattlegroundsQuestRewardEquippedParser(ParserState ParserState, StateFacade facade)
        {
            this.ParserState = ParserState;
            this.GameState = ParserState.GameState;
            this.StateFacade = facade;
        }

        public bool AppliesOnNewNode(Node node, StateType stateType)
        {
            return stateType == StateType.PowerTaskList
                && node.Type == typeof(TagChange)
                && (node.Object as TagChange).Name == (int)GameTag.ZONE
                && (node.Object as TagChange).Value == (int)Zone.PLAY
                && GameState.CurrentEntities.ContainsKey((node.Object as TagChange).Entity)
                && GameState.CurrentEntities[(node.Object as TagChange).Entity].GetTag(GameTag.CARDTYPE) == (int)CardType.BATTLEGROUND_QUEST_REWARD;
        }

        public bool AppliesOnCloseNode(Node node, StateType stateType)
        {
            return stateType == StateType.PowerTaskList
                && node.Type == typeof(FullEntity)
                && (node.Object as FullEntity).GetTag(GameTag.ZONE) == (int)Zone.PLAY
                && (node.Object as FullEntity).GetTag(GameTag.CARDTYPE) == (int)CardType.BATTLEGROUND_QUEST_REWARD;
        }

        public List<GameEventProvider> CreateGameEventProviderFromNew(Node node)
        {
            var tagChange = node.Object as TagChange;
            var entity = GameState.CurrentEntities.GetValueOrDefault(tagChange.Entity);
            if (entity == null)
            {
                return null;
            }

            var cardId = entity.CardId;
            var controllerId = entity.GetEffectiveController();
            var isHeroPowerReward = entity.GetTag(GameTag.BACON_IS_HEROPOWER_QUESTREWARD) == 1;
            return new List<GameEventProvider> { GameEventProvider.Create(
                tagChange.TimeStamp,
                "BATTLEGROUNDS_QUEST_REWARD_EQUIPPED",
                GameEvent.CreateProvider(
                    "BATTLEGROUNDS_QUEST_REWARD_EQUIPPED",
                    cardId,
                    controllerId,
                    entity.Id,
                    StateFacade,
                    //null,
                    new {
                        PlayerId = entity.GetTag(GameTag.PLAYER_ID),
                        IsHeroPowerReward = isHeroPowerReward,
                    }
                ),
                true,
                node) };
        }

        public List<GameEventProvider> CreateGameEventProviderFromClose(Node node)
        {
            var fullEntity = node.Object as FullEntity;
            var cardId = fullEntity.CardId;
            var controllerId = fullEntity.GetEffectiveController();
            var isHeroPowerReward = fullEntity.GetTag(GameTag.BACON_IS_HEROPOWER_QUESTREWARD) == 1;
            return new List<GameEventProvider> { GameEventProvider.Create(
                fullEntity.TimeStamp,
                "BATTLEGROUNDS_QUEST_REWARD_EQUIPPED",
                GameEvent.CreateProvider(
                    "BATTLEGROUNDS_QUEST_REWARD_EQUIPPED",
                    cardId,
                    controllerId,
                    fullEntity.Id,
                    StateFacade,
                    //null,
                    new {
                        PlayerId = fullEntity.GetTag(GameTag.PLAYER_ID),
                        IsHeroPowerReward = isHeroPowerReward,
                    }
                ),
                true,
                node) };
        }
    }
}
